### PR TITLE
fix: add hatchling for platformdirs

### PR DESCRIPTION
**Pull request by yajo**
https://github.com/platformdirs/platformdirs/pull/75 was released yesterday.

### CODE REVIEW

**Comment by Yajo** [comment]
Any clues on why this is failing?

**Comment by takeda** [comment]
@Yajo This is just guessing, I'm not as familiar with poetry2nix internals, but I'm wondering if it isn't due to pyproject.toml not listing any dependencies, yet, the wheel on PyPI lists these (I'm thinking poetry2nix compiles dependencies it uses pyproject.toml?):

```
Requires-Dist: editables>=0.2; python_version > '3'
Requires-Dist: importlib-metadata; python_version < '3.8'
Requires-Dist: packaging>=21.3; python_version > '3'
Requires-Dist: packaging~=20.9; python_version < '3'
Requires-Dist: pathspec>=0.9
Requires-Dist: pluggy>=1.0.0; python_version > '3'
Requires-Dist: pluggy~=0.13; python_version < '3'
Requires-Dist: tomli>=1.2.2; python_version > '3'
Requires-Dist: toml~=0.10.2; python_version < '3'
```

Perhaps an override for hatchling is also necessary?

@ofek why the dependencies are not in pyproject.toml?

File: overrides/build-systems.json [hunk 1]
```diff
@@ -833,6 +833,10 @@
   "pixelmatch": [
     "poetry-core"
   ],
+  "platformdirs": [
+    "hatchling",
+    "packaging"
```
Comment (takeda) [inline overrides/build-systems.json:838]: This file only contains the build system used. So it's good place to list hatchling for platformdirs, but not this kind of stuff.

I was thinking something more like adding to `overrides/default.nix`:

```nix
hatchling = super.hatchling.overridePythonAttrs (
  old: {
    propagatedBuildInputs = (old.propagatedBuildInputs or [ ]) ++ [
      self.editables
      self.packaging
      self.pathspec
      self.pluggy
      self.tomli
    ] ++ lib.optionals (self.pythonOlder "3.8") [
      self.importlib-metadata
    ];
  }
);
```

**Comment by adisbladis** [comment]
When we add something via an override that isn't in the projects `poetry.lock` we inherit whatever version that nixpkgs has.
In this instance `poetry` has a dependency on `platformdirs` which pulled in `hatchling` via the override.

The problem was that Poetry's dependency on packaging was defined as `packaging = "^20.4"` while hatchling requires 21.3.

For now I've added a conditional around this override that just checks if platformdirs is really a version that requires hatchling.
This still isn't technically correct, but it could certainly be worse.

**Review by yajo — COMMENTED** [review]
Seems good.

**Comment by xanderio** [comment]
Even with this fix applied platformdirs still fails to build.

```
Processing /build/platformdirs-2.5.2
  Running command Preparing metadata (pyproject.toml)
  Traceback (most recent call last):
    File "/nix/store/ycgq5lbkywhrwr8gmyj4wjbdzx0cf4b7-python3.8-pip-22.0.3/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 156, in prepare_metadata_for_build_wheel
      hook = backend.prepare_metadata_for_build_wheel
  AttributeError: module 'hatchling.build' has no attribute 'prepare_metadata_for_build_wheel'

  During handling of the above exception, another exception occurred:

  Traceback (most recent call last):
    File "/nix/store/ycgq5lbkywhrwr8gmyj4wjbdzx0cf4b7-python3.8-pip-22.0.3/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 363, in <module>
      main()
    File "/nix/store/ycgq5lbkywhrwr8gmyj4wjbdzx0cf4b7-python3.8-pip-22.0.3/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 345, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/nix/store/ycgq5lbkywhrwr8gmyj4wjbdzx0cf4b7-python3.8-pip-22.0.3/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 160, in prepare_metadata_for_build_wheel
      whl_basename = backend.build_wheel(metadata_directory, config_settings)
    File "/nix/store/lpssqjfy8c05nmwav954w1n9531nlzn4-python3.8-hatchling-0.22.0/lib/python3.8/site-packages/hatchling/build.py", line 41, in build_wheel
      return os.path.basename(next(builder.build(wheel_directory, ['standard'])))
    File "/nix/store/lpssqjfy8c05nmwav954w1n9531nlzn4-python3.8-hatchling-0.22.0/lib/python3.8/site-packages/hatchling/builders/plugin/interface.py", line 103, in build
      configured_build_hooks = self.get_build_hooks(directory)
    File "/nix/store/lpssqjfy8c05nmwav954w1n9531nlzn4-python3.8-hatchling-0.22.0/lib/python3.8/site-packages/hatchling/builders/plugin/interface.py", line 318, in get_build_hooks
      raise ValueError('Unknown build hook: {}'.format(hook_name))
  ValueError: Unknown build hook: vcs
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /nix/store/4g4p6h9gv9r44w261vsn5k2wc905pqhw-python3-3.8.13/bin/python3.8 /nix/store/ycgq5lbkywhrwr8gmyj4wjbdzx0cf4b7-python3.8-pip-22.0.3/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /build/tmp4h5n1fto
  cwd: /build/platformdirs-2.5.2
  Preparing metadata (pyproject.toml) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

**Comment by xanderio** [comment]
Even with this fix applied platformdirs still fails to build.

```
Processing /build/platformdirs-2.5.2
  Running command Preparing metadata (pyproject.toml)
  Traceback (most recent call last):
    File "/nix/store/ycgq5lbkywhrwr8gmyj4wjbdzx0cf4b7-python3.8-pip-22.0.3/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 156, in prepare_metadata_for_build_wheel
      hook = backend.prepare_metadata_for_build_wheel
  AttributeError: module 'hatchling.build' has no attribute 'prepare_metadata_for_build_wheel'

  During handling of the above exception, another exception occurred:

  Traceback (most recent call last):
    File "/nix/store/ycgq5lbkywhrwr8gmyj4wjbdzx0cf4b7-python3.8-pip-22.0.3/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 363, in <module>
      main()
    File "/nix/store/ycgq5lbkywhrwr8gmyj4wjbdzx0cf4b7-python3.8-pip-22.0.3/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 345, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/nix/store/ycgq5lbkywhrwr8gmyj4wjbdzx0cf4b7-python3.8-pip-22.0.3/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 160, in prepare_metadata_for_build_wheel
      whl_basename = backend.build_wheel(metadata_directory, config_settings)
    File "/nix/store/lpssqjfy8c05nmwav954w1n9531nlzn4-python3.8-hatchling-0.22.0/lib/python3.8/site-packages/hatchling/build.py", line 41, in build_wheel
      return os.path.basename(next(builder.build(wheel_directory, ['standard'])))
    File "/nix/store/lpssqjfy8c05nmwav954w1n9531nlzn4-python3.8-hatchling-0.22.0/lib/python3.8/site-packages/hatchling/builders/plugin/interface.py", line 103, in build
      configured_build_hooks = self.get_build_hooks(directory)
    File "/nix/store/lpssqjfy8c05nmwav954w1n9531nlzn4-python3.8-hatchling-0.22.0/lib/python3.8/site-packages/hatchling/builders/plugin/interface.py", line 318, in get_build_hooks
      raise ValueError('Unknown build hook: {}'.format(hook_name))
  ValueError: Unknown build hook: vcs
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /nix/store/4g4p6h9gv9r44w261vsn5k2wc905pqhw-python3-3.8.13/bin/python3.8 /nix/store/ycgq5lbkywhrwr8gmyj4wjbdzx0cf4b7-python3.8-pip-22.0.3/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /build/tmp4h5n1fto
  cwd: /build/platformdirs-2.5.2
  Preparing metadata (pyproject.toml) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

**Comment by lelit** [comment]
> Even with this fix applied platformdirs still fails to build.

Right, got the same yesterday, adding just `hatchling`. [This comment](https://github.com/platformdirs/platformdirs/commit/d6ab2ca5bb19d0c707352d826d76785d085d165f#commitcomment-71548711) suggests that `hatch-vcs` is needed too, but I could not try that out yet.

**Comment by lelit** [comment]
See also [this other comment](https://github.com/ofek/hatch/issues/128#issuecomment-1106537506).

**Comment by adisbladis** [comment]
I've added `hatch-vcs` in 8cfd980262181bd3ef15899708ceeb2e3f33958b.
It's _still_ not going to work unless you add `hatch-vcs` to your dependencies at the moment since `hatch-vcs` isn't packaged in nixpkgs, and as such we cannot inherit it from there.

I don't have the mental bandwidth to package this for nixpkgs, but feel free to cc me on a nixpkgs PR for a quick review.

**Pull request merged**